File: frontend/src/api/auditLog.js
/**
 * Audit Log API client
 *
 * Provides functions to interact with the audit log endpoints
 * for viewing edit history and performing moderation actions.
 */
import apiClient from './client';

export const auditLogApi = {
    /**
     * Get list of edits with optional filters
     * @param {Object} params - Query parameters
     * @param {string[]} params.status - Filter by status(es): PENDING, APPROVED, REJECTED, REVERTED
     * @param {string} params.entity_type - Filter by entity type
     * @param {string} params.user_id - Filter by submitter user ID
     * @param {string} params.start_date - Filter by start date (ISO 8601)
     * @param {string} params.end_date - Filter by end date (ISO 8601)
     * @param {number} params.skip - Pagination offset
     * @param {number} params.limit - Pagination limit
     */
    getList: (params = {}) =>
        apiClient.get('/api/v1/audit-log', { params }),

    /**
     * Get full details of a single edit
     * @param {string} editId - UUID of the edit
     */
    getDetail: (editId) =>
        apiClient.get(`/api/v1/audit-log/${editId}`),

    /**
     * Get count of pending edits (for notification badge)
     */
    getPendingCount: () =>
        apiClient.get('/api/v1/audit-log/pending-count'),

    /**
     * Revert an approved edit
     * @param {string} editId - UUID of the edit to revert
     * @param {Object} data - Request body with optional notes
     */
    revert: (editId, data = {}) =>
        apiClient.post(`/api/v1/audit-log/${editId}/revert`, data),

    /**
     * Re-apply a reverted or rejected edit
     * @param {string} editId - UUID of the edit to re-apply
     * @param {Object} data - Request body with optional notes
     */
    reapply: (editId, data = {}) =>
        apiClient.post(`/api/v1/audit-log/${editId}/reapply`, data),
};

export default auditLogApi;

File: backend/app/schemas/audit_log.py
"""
Pydantic schemas for the Audit Log feature.

These schemas define the API request/response models for the audit log endpoints,
which provide both moderation queue and audit history functionality.
"""
from pydantic import BaseModel, ConfigDict
from typing import Optional, Dict, Any, List
from datetime import datetime, date


class UserSummary(BaseModel):
    """Minimal user information for display in audit log entries."""
    user_id: str
    display_name: Optional[str]
    email: str

    model_config = ConfigDict(from_attributes=True)


class AuditLogEntryResponse(BaseModel):
    """
    Summary of an edit for the audit log list view.
    
    Contains all columns needed for the grid display, with human-readable
    entity names (not UUIDs).
    """
    edit_id: str
    status: str  # PENDING, APPROVED, REJECTED, REVERTED
    entity_type: str  # TEAM, ERA, SPONSOR, BRAND, SPONSOR_LINK, LINEAGE
    entity_name: str  # Human-readable name (resolved from UUID)
    action: str  # CREATE, UPDATE, DELETE
    submitted_by: UserSummary
    submitted_at: datetime
    reviewed_by: Optional[UserSummary] = None
    reviewed_at: Optional[datetime] = None
    summary: str  # Reason → Internal Note → Rejection notes (priority)

    model_config = ConfigDict(from_attributes=True)


class AuditLogListResponse(BaseModel):
    items: List[AuditLogEntryResponse]
    total: int



class AuditLogDetailResponse(BaseModel):
    """
    Full details of an edit for the detail/diff view.
    
    Includes before/after snapshots with human-readable names and
    permission flags indicating what actions the current user can take.
    """
    edit_id: str
    status: str
    entity_type: str
    entity_name: str
    action: str
    submitted_by: UserSummary
    submitted_at: datetime
    reviewed_by: Optional[UserSummary] = None
    reviewed_at: Optional[datetime] = None
    summary: str
    # Detailed snapshots with human-readable names
    snapshot_before: Optional[Dict[str, Any]] = None
    snapshot_after: Dict[str, Any]
    source_url: Optional[str] = None
    source_notes: Optional[str] = None
    review_notes: Optional[str] = None
    # Permission flags for current user
    can_approve: bool
    can_reject: bool
    can_revert: bool  # Only if most recent approved and user has permission
    can_reapply: bool  # Only if reverted/rejected and chronologically valid

    model_config = ConfigDict(from_attributes=True)


class AuditLogFilters(BaseModel):
    """Query parameters for filtering the audit log list."""
    status: Optional[List[str]] = None  # Default handled in endpoint
    entity_type: Optional[str] = None
    user_id: Optional[str] = None  # Filter by submitter
    entity_id: Optional[str] = None  # Filter by specific entity
    entity_search: Optional[str] = None  # Search by entity name
    date_from: Optional[date] = None
    date_to: Optional[date] = None


class ReviewEditRequest(BaseModel):
    """Request body for approving or rejecting an edit."""
    approved: bool
    notes: Optional[str] = None


class ReviewEditResponse(BaseModel):
    """Response after reviewing an edit."""
    edit_id: str
    status: str
    message: str


class RevertRequest(BaseModel):
    """Request body for reverting an approved edit."""
    notes: Optional[str] = None


class ReapplyRequest(BaseModel):
    """Request body for re-applying a reverted or rejected edit."""
    notes: Optional[str] = None


class AuditLogStatsResponse(BaseModel):
    """Statistics for the audit log dashboard."""
    pending_count: int
    approved_today: int
    rejected_today: int
    reverted_today: int
    pending_by_type: Dict[str, int]
